examples/optimizers-cpp/Main.hs

```diff
@@ -0,0 +1,134 @@
+{-# LANGUAGE BlockArguments #-}
+
+module Main where
+
+{- Optimizer implementations run on standard test functions -}
+
+import Control.Monad (foldM, when, forM_)
+import GHC.Generics
+import Prelude hiding (sqrt)
+import Text.Printf (printf)
+import Data.Default.Class
+
+import Torch hiding (Optimizer(..), runStep, Adam(..))
+import TestFunctions
+import Torch.Optim.Internal
+
+-- import Data.IORef
+
+-- | show output after n iterations
+showLog :: (Show a) => Int -> Int -> Int -> Tensor -> a -> IO ()
+showLog n i maxIter lossValue state = 
+    when (i == 0 || mod i n == 0 || i == maxIter-1) do
+        putStrLn ("Iter: " ++ printf "%6d" i
+            ++ " | Loss:" ++ printf "%05.4f" (asValue lossValue :: Float)
+            ++ " | Parameters: " ++ show state)
+
+-- | Optimize convex quadratic with specified optimizer
+optConvQuad :: (Optimizer opt) => Int -> opt -> IO ()
+optConvQuad numIter optInit = do
+    let dim = 2
+        a = eye' dim dim
+        b = zeros' [dim]
+    paramInit <- sample $ ConvQuadSpec dim
+    putStrLn ("Initial :" ++ show paramInit)
+    optimizer <- initOptimizer optInit paramInit
+    forM_ [1..numIter] \i -> do
+      step optimizer $ \paramState -> do
+        let lossValue = (lossConvQuad a b) paramState
+        showLog 1000 i numIter lossValue paramState
+        return lossValue
+    trained <- getParams optimizer :: IO ConvQuad
```

Comment:
? the model parameters are coming from the optimizer?

Reply:
Yes.
The optimizer has model parameters, optimizer state and optimizer option.
https://github.com/pytorch/pytorch/blob/1.6/torch/csrc/api/include/torch/optim/optimizer.h#L147-L149